packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -562,6 +562,10 @@ function recursivelyDisconnectChildren(vnodes: VNodes) {
 export function resetShadowRoot(vm: VM) {
     const { children, cmpRoot, renderer } = vm;
 
+    if (!hasShadow(vm)) {
+        return;
+    }
+
```

Comment:
I'm sure there's a way to do this performance optimization for light DOM as well. Happy to add a TODO if that makes sense.

Reply:
Hm, after noticing this comment, I think we actually do need to solve this for light DOM:

https://github.com/salesforce/lwc/blob/28e715b92caefa69aa83b510a45eb494715c12d8/packages/%40lwc/engine-core/src/framework/template.ts#L132-L134